### ISSUE TITLE
jsmn_util.c - do you want extras?

### ISSUE THREAD

**Issue by zserge** [zserge/jsmn#48]
Thanks for the parser, surprisingly easy to use!

I found I wanted some extra stuff, like unescaping strings and converting jsmntype_t into text.  I wrote some stuff (it's in a Git repo on another project - cf. issue #47) and with tests, so I could convert back to .hg and make a pull request.

Do you want extras, given they could be completely omitted and not increase the footprint if there were unwanted?

What about additions that add a (void*) to jsmn_parser and some extra code to jsmn.c, for generating errors?

---
- Bitbucket: https://bitbucket.org/zserge/jsmn/issue/48
- Originally reported by: Matthew Astley
- Originally created at: 2015-05-13T18:14:11.737

**Comment by zserge** [comment]
https://github.com/mcast/json-fileops/tree/jsmn/reHGify

---

Original comment by: Matthew Astley

**Comment by zserge** [comment]
/\* Wide character _/
/_ UTF8 not yet implemented.  Pretend we didn't see it. */

---

Original comment by: Gabriel Gritsch

**Comment by zserge** [comment]
@MacGritsch True, I didn't make an escape hatch here for errors.  It translates as in [new jsmn_test.c fn](https://github.com/mcast/json-fileops/blob/a662a7d6269b8dde1ef23dcecca00c8989589fb0/jsmn_test.c#L609-L615),

`"\\u0000\\u00ff" --> "\x00\xff"` but
`"\\u0100" --> "\\u0100"` (unchanged).

The plumbing is there, it just needs a  [u8_wc_toutf8](http://www.cprogramming.com/tutorial/unicode.html) or similar, and the relevant bounds checking.

My data doesn't need it...  I'm not going to venture an answer on whether "minimal" is compatible with "Unicode".  I do think 70% of users parsing strings are going to want them unescaped though.

---

Original comment by: Matthew Astley

**Comment by zserge** [comment]
Let's vote. I personally don't need such an extension, but if JSMN users find it useful - we can add it either as a wrapper on top of jsmn or something similar.
My feeling is that JSMN is one of the most "user-hostile" parsers, but its memory efficiency and speed make it a winner on embedded systems and that's the target audience of JSMN.
Would be happy to hear vox populi!

**Comment by eNTi** [comment]
i want it... almost every non-english speaking country will want to have some form of unicode handling. not doing it in your lib means everyone (who needs this) will have to put in their code base regardless. i suspect it is more than just a handful of people and it is probably even a criteria over which this lib is picked or not. if i knew that before hand i would have chosen another library. at least make it very clear on the front page that you do not handle utf8 at all.

**Comment by legitparty** [comment]
My fork has Unicode, a full DOM, and an emitter. Can use libc or its own methods for the string formatting, and has its own UTF-8 and JSON string code, so no need for a big Unicode library. Check it out if you need anything. It is all contributed back with the same license. I just haven't gotten around to documenting it properly, but it is all pretty obvious to someone familiar with JSMN.

**Comment by schneidersoft** [comment]
Another take on json string decoding (untested, just a rough idea):
https://gist.github.com/schneidersoft/ca11078da9e0607501b85a0cdef317bc

**Comment by pt300** [comment]
@schneidersoft your solution looks to be vulnerable to out of bounds reads. 

I believe code like this would be good for some kind of jsmn extra utilities. JSMN is still focused at memory constrained systems and not every use of jsmn will even need decoding of unicode, but for those who need it additional header file could be added.

Guess this goes goes into my bucket of things to properly analyse and do.
